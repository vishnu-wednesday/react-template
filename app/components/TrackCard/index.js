/**
 *
 * TrackCard
 *
 */
import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Card, Avatar, Skeleton } from 'antd';
import { injectIntl } from 'react-intl';
import { compose } from 'redux';
import { Link } from 'react-router-dom';

import { colors, fonts, styles } from '@themes';
import T from '@components/T';
import If from '@components/If';

const { Meta } = Card;

const Audio = styled.audio`
  width: 100%;
`;

const StyleT = styled(T)`
  && {
    color: ${colors.artistNameColor};
    ${fonts.style.standard()};
  }
`;

const AlbumNameT = styled(T)`
  && {
    color: ${colors.artistNameColor};
    ${fonts.style.subheading()};
  }
`;

const StyledCaption = styled.figcaption`
  clear: both;
  ${styles.margin.top(1)}
  ${styles.margin.left(0.5)}
`;

export function TrackCard({ index, track, intl, loading }) {
  const getTrackOrCollection = (item) => {
    if (item.wrapperType === 'track') {
      return {
        name: item.trackName,
        id: item.trackId
      };
    }
    return {
      name: item.collectionName,
      id: item.collectionId
    };
  };

  return (
    <>
      <If condition={!!track}>
        <Link to={`/${getTrackOrCollection(track).id}`} key={index}>
          <Card key={index} data-testid="track-card">
            <Skeleton loading={loading} avatar active>
              <Meta
                avatar={
                  <Avatar
                    size={{
                      xs: 32,
                      sm: 38,
                      md: 40,
                      lg: 64,
                      xl: 80,
                      xxl: 100
                    }}
                    src={track.artworkUrl100}
                  />
                }
                description={
                  <div>
                    <AlbumNameT
                      id="track_or_collection"
                      values={{ value: getTrackOrCollection(track).name }}
                      clearFloat={true}
                      data-testid="track_or_collection"
                    />
                    <StyleT id="artist_name" values={{ artistName: track.artistName }} clearFloat={true} />
                  </div>
                }
              />
              <figure>
                <StyledCaption>{intl.formatMessage({ id: 'listen_to_it' })}</StyledCaption>
                <Audio controls src={track.previewUrl} data-testid="audio-player">
                  <T id="audio_no_support" values={{ code: (chunks) => <code>{chunks}</code> }} />
                </Audio>
              </figure>
            </Skeleton>
          </Card>
        </Link>
      </If>
    </>
  );
}

TrackCard.propTypes = {
  index: PropTypes.number,
  track: PropTypes.object,
  intl: PropTypes.object,
  loading: PropTypes.bool,
  history: PropTypes.object
};

TrackCard.defaultProps = {
  track: {}
};

export default compose(injectIntl)(TrackCard);
