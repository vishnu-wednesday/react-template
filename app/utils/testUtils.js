import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { browserHistory, BrowserRouter, MemoryRouter, Route } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import { createIntl, createIntlCache, IntlProvider } from 'react-intl';

import configureStore from '@app/configureStore';
import { DEFAULT_LOCALE, translationMessages } from '@app/i18n';
import ConnectedLanguageProvider from '@containers/LanguageProvider';
import { IntlGlobalProvider } from '@components/IntlGlobalProvider';
import { createMemoryHistory } from 'history';

export const renderWithIntl = (children) =>
  render(
    <IntlProvider locale={DEFAULT_LOCALE} messages={translationMessages[DEFAULT_LOCALE]}>
      <BrowserRouter>
        <IntlGlobalProvider>{children}</IntlGlobalProvider>
      </BrowserRouter>
    </IntlProvider>
  );

export const getComponentStyles = (Component, props = {}) => {
  renderWithIntl(Component(props));
  const { styledComponentId } = Component(props).type;
  const componentRoots = document.getElementsByClassName(styledComponentId);
  // eslint-disable-next-line no-underscore-dangle
  return window.getComputedStyle(componentRoots[0])._values;
};

export const renderProvider = (children) => {
  const store = configureStore({}, browserHistory).store;
  return render(
    <Provider store={store}>
      <ConnectedLanguageProvider messages={translationMessages}>
        <ThemeProvider
          theme={{
            main: 'violet'
          }}
        >
          {children}
        </ThemeProvider>
      </ConnectedLanguageProvider>
    </Provider>
  );
};

export function renderWithRouterMatch(
  ui,
  { path = '/', route = '/', history = createMemoryHistory({ initialEntries: [route] }) } = {}
) {
  const store = configureStore({}, browserHistory).store;
  return render(
    <Provider store={store}>
      <IntlProvider locale={DEFAULT_LOCALE} messages={translationMessages[DEFAULT_LOCALE]}>
        <MemoryRouter initialEntries={[route]}>
          <Route path={path}>{ui}</Route>
        </MemoryRouter>
      </IntlProvider>
    </Provider>
  );
}

export const timeout = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
export const apiResponseGenerator = (ok, data) => ({
  ok,
  data
});

export const createIntlUtil = (locale = DEFAULT_LOCALE) => {
  const cache = createIntlCache();
  const getMessages = (locale) => translationMessages[locale];

  return createIntl(
    {
      locale: locale,
      messages: getMessages(locale)
    },
    cache
  );
};
