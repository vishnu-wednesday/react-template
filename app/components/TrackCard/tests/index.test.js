/**
 *
 * Tests for TrackCard
 *
 */

import React from 'react';
// import { fireEvent } from '@testing-library/dom'
import { renderWithIntl } from '@utils/testUtils';
import TrackCard from '../index';
import { mockData } from './mockTrackResponse';

describe('<TrackCard />', () => {
  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<TrackCard track={{}} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should render one card component', () => {
    const { getByTestId } = renderWithIntl(<TrackCard track={{}} />);
    expect(getByTestId('track-card')).toBeInTheDocument();
  });

  it('should show trackName for type track ', () => {
    const { queryByText } = renderWithIntl(<TrackCard track={mockData.results[0]} />);
    // eslint-disable-next-line no-console
    expect(queryByText(mockData.results[0].trackName)).toBeDefined();
    expect(queryByText(mockData.results[0].collectionName)).toBeNull();
  });

  it('should show collectionName for type collection ', () => {
    const { queryByText } = renderWithIntl(<TrackCard track={mockData.results[1]} />);
    // eslint-disable-next-line no-console
    expect(queryByText(mockData.results[1].trackName)).toBeNull();
    expect(queryByText(mockData.results[1].collectionName)).toBeDefined();
  });
});
