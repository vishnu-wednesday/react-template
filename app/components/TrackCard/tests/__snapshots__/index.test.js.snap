// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackCard /> should render and match the snapshot 1`] = `
<body>
  <div>
    <a
      href="/track/undefined"
    >
      <div
        class="ant-card ant-card-bordered"
        data-testid="track-card"
      >
        <div
          class="ant-card-body"
        >
          <div
            class="ant-card-meta"
          >
            <div
              class="ant-card-meta-avatar"
            >
              <span
                class="ant-avatar ant-avatar-circle"
              >
                <span
                  class="ant-avatar-string"
                  style="opacity: 0;"
                />
              </span>
            </div>
            <div
              class="ant-card-meta-detail"
            >
              <div
                class="ant-card-meta-description"
              >
                <div>
                  <p
                    class="T__StyledText-gjlic1-0 bxtGdb TrackCard__AlbumNameT-sc-10bk196-2 hhggcZ"
                    data-testid="track_or_collection"
                  >
                    
                  </p>
                  <p
                    class="T__StyledText-gjlic1-0 bxtGdb TrackCard__StyleT-sc-10bk196-1 hZFoBe"
                    data-testid="t"
                  >
                    
                  </p>
                </div>
              </div>
            </div>
          </div>
          <figure>
            <figcaption
              class="TrackCard__StyledCaption-sc-10bk196-3 jhwGow"
            >
              Play it!
            </figcaption>
            <audio
              class="TrackCard__Audio-sc-10bk196-0 ipQWAE"
              controls=""
              data-testid="audio-player"
            >
              <p
                class="T__StyledText-gjlic1-0 jbStwN"
                data-testid="t"
              >
                Your browser doesn't support 
                <code>
                  audio
                </code>
                 HTML element
              </p>
            </audio>
          </figure>
        </div>
      </div>
    </a>
  </div>
</body>
`;
